wip: string overhaul

StefanKarpinski opened the pull request:
Making a PR to run CI.

Keno commented on base/iostream.jl:
```diff
@@ -336,12 +336,12 @@ function read(s::IOStream, nb::Integer; all::Bool=true)
 
 ## Character streams ##
-const _chtmp = Ref{Char}()
+const _chtmp = Ref{UInt32}()
```
Unrelated, but this should now be inside the function :)

bkamins commented:
Nice! For testing purposes I have written a verbose version of the decoder here https://gist.github.com/bkamins/5a2f89cab14d434e3f3e4a23c121a8a2 that checks for validity of UTF-8 (it is slower of course).
I have checked and both implementations return the same values for 1112064 valid UTF-8.

bkamins commented on base/char.jl:
```diff
@@ -1,8 +1,24 @@
+    ((u & 0x00ff0000) >> 4) ⊻ ((u & 0xff000000) >> 6)
+end
+
+function convert(::Type{Char}, u::UInt32)
```
Tested and this works OK for all valid Unicode (last code point U+10FFFF). The conversion breaks for invalid Unicode as it overflows at `0x00400000`.
I understand that we want to avoid branching here so adding the test if `u` is not too large is not an option? If so - this probably be documented somewhere.

StefanKarpinski replied:
I'm still contemplating how best to handle invalid code points. There was previously an error check here but it caused a massive build time regression (it was also calling a deprecated error type), so I just took it out for now. Will have to figure out what to do here.

bkamins commented:
With respect to validity checking:
Neither `convert(::Type{UInt32}, c::Char)` nor `convert(::Type{Char}, u::UInt32)` are injective - different arguments (fortunately invalid in cases of both functions) are mapped to the same value.
Thus `convert(UInt32, convert(Char, u))` does not have to be equal to `u` if `u` is invalid, similarly for `convert(Char, convert(UInt32, c))` if `c` is invalid.
In this light maybe something similar to `@inbounds` macro might be a good idea (turning checking if argument is valid on and off)?

StefanKarpinski commented:
Either that or change the transformations to be bijective. I’m considering how to do that.

bkamins commented:
My thoughts on the bijection (I hope there is no error in this 😄, but I have tested it not only in the head but also by implementation).

The best I could think of it is doable but a bit messy (fortunately only for invalid values). Here is the outline how this can be done (giving description of mapping `UTF-8` to `UInt32`):
* assume that all 3-byte `ED` sequences are valid (only for bijection in conversion - not for validity testing; this simplifies things)
* decode all valid `UTF-8` byte sequences in `Char` to their `UInt32` representation;
* all invalid `UTF-8` byte sequences in `Char` that have representation larger than `0x10FFFF` should be left unaltered (so we apply identity function to them - they were invalid in UTF-8 and remain invalid in `UInt32` representation);
* all invalid `UTF-8` byte sequences in `Char` that have representation less than `0x10FFFF` should be mapped to representation of valid `UTF-8` byte sequences in `Char` that have representation larger than `0x10FFFF`.

There are exactly 1112064 valid `Char` greater than `0x10FFFF` (call them large valid) and the same number of invalid `Char` less than `0x10FFFF` (call them small invalid). Now the trick is that most of small invalid `Char` are equal to decoded large valid `Char` so the conversion is simple.
Only 1920 small invalid values in the range from 128 to 2047 are not directly mapped to large valid `Char` (as they are encoded by two byte small valid `Char`). And those have to be mapped to large valid `Char` that when decoded produce two-byte small valid `Char` (again there are 1920 such values and they are pretty simple to indentify).

bkamins commented on base/filesystem.jl:
```diff
@@ -149,6 +149,26 @@ function read(f::File, ::Type{UInt8})
+        while 1 < n
+            p = position(f)
+            b = read(f, UInt8)
+            if b & 0xc0 != 0x80
```
Here we should also handle sequences starting with `0xC0`, `0xC1`, `0xE0`, `0xF0` that would be overlong, starting with `0xED` that would be reserved and starting with `0xF4` that would encode too large code point.

StefanKarpinski replied:
I'm not sure about that. I think that parsing overlong encodings as a single char is a much friendlier behavior. They occur fairly often in real data, many systems (e.g. [Java](https://en.wikipedia.org/wiki/UTF-8#Modified_UTF-8)) use overlong encodings to avoid embedded nul bytes.

bkamins commented on base/io.jl:
```diff
@@ -462,22 +451,19 @@ function read!(s::IO, a::Array{T}) where T
+    c = UInt32(b0)
+    if n <= 4
+        while 1 < n
+            peek(s) & 0xc0 == 0x80 || break
```
The same comment as above regarding handling invalid sequences.

bkamins commented on base/iostream.jl:
```diff
@@ -336,14 +336,23 @@ function read(s::IOStream, nb::Integer; all::Bool=true)
 function peekchar(s::IOStream)
-    if ccall(:ios_peekutf8, Cint, (Ptr{Void}, Ptr{Char}), s, _chtmp) < 0
+    chref = Ref{UInt32}()
+    if ccall(:ios_peekutf8, Cint, (Ptr{Void}, Ptr{UInt32}), s, chref) < 0
```
current implementation of `ios_peekutf8` also does not check for invalid sequences.

stevengj replied:
Seems a bit wasteful for `ios_peekutf8` to convert the UTF-8 encoding to `UInt32`, and then to convert it back when returning `Char(chref[])`.

Since this seems to be the only function in all of Julia that calls `ios_peekutf8`, we should just re-write `ios_peekutf8` to return `Char`.

StefanKarpinski replied:
Agree, this stuff should be optimized to avoid converting back and forth. This PR is WIP and the first step is just to get everything working and all tests passing.

StefanKarpinski replied:
This is still in my new PR, which is annoying but I think that all the `ios_` support code needs an overhaul in the 1.x timeframe and ideally we would move most of the buffering code etc. into Julia and it can understand the new `Char` representation and avoid the inefficiency here. Until then I think it's ok to just leave this as it is.

bkamins commented:
In the above comments I have followed strict Unicode 10 compliant approach to reading of UTF-8.
However, it could be OK to create invalid `Char` starting with `0xC0`, `0xC1`, `0xE0`, `0xF0`, `0xED`, `0xF4` trying to read the expected number of bytes provided that later we check that they are invalid.

The reason is that if they were created they would have `10` leading bits in all continuation bytes, so there is no risk of doing this as such following byte would be invalid anyway.

This will probably speed up `Char` creation. Of course then it is even more important to precisely check `Char` for validity.

If this approach is taken the only issue to fix is `ios_peekutf8` which sometimes tries to read 4 or 5 continuation bytes (as specified in `trailingBytesForUTF8`), and this should not be allowed (we should read up to 3 continuation bytes as otherwise they will not fit into `UInt32`).

bkamins commented:
As an additional comment to keep in mind - if we use strict parsing rule we are then sure that e.g. string length is consistent with Unicode 10 compliant parsers (as @StefanKarpinski mentioned earlier). If we use simplified `Char` creation we lose this property.

StefanKarpinski commented:
We should definitely not ever read 4-5 continuation bytes; that's from a very old UTF-8 specification and hasn't been allowed for a long time. It's also not a thing that happens much in real life (aside from CESU-8 maybe?), and would require a larger `Char` type, which is undesirable.

StefanKarpinski commented:
See also WTF-8: https://en.wikipedia.org/wiki/UTF-8#WTF-8, which ignores invalidity of surrogate pairs, since UTF-8 has no intrinsic issue with them – they are purely invalid because of how UTF-16 works (and it only works that way because of its legacy history with UCS-2).

StefanKarpinski commented:
Some cases I can see for invalid UTF-8 data, together with the most useful "character" parsing:

1. Binary data: `Char` = byte
2. Latin-1 data: `Char` = byte
3. Modified UTF-8: `Char` = loosely decoded UTF-8 characters
4. WTF-8: `Char` = loosely decoded UTF-8 characters
5. Mangled UTF-8 data: `Char` = partial characters?

It's hard to imagine where the Unicode 10 partial character decoding business is actually a useful interpretation. Of course, the byte interpretation for binary and Latin-1 data isn't really very useful either since you don't get the right characters out, but at least the number and size of the units are correct. And yes, there's a chance that binary data or Latin-1 will look like valid UTF-8, but that's very improbable, especially if you require a complete UTF-8 character to group bytes. So I guess that in contrast to Unicode 10, I think the most useful `Char` splitting approach would be:

* Split anything that has the basic form of a UTF-8 character as a `Char`: i.e. an ASCII byte or a lead byte with 2-4 leading ones followed by that many continuation bytes. This covers ASCII, valid UTF-8 (which includes ASCII), Modified UTF-8, and WTF-8. I suspect we should also just allow converting these to integer code point values in the straightforward way – some of which will produce code points of surrogate pairs (which one can then check for validity) and some of which will map to the same code points, non-injectively (i.e. overlong encodings). On the other hand, two different encodings of the same code point should not be equal as characters. So, for example, `c = '\xC0\x80'` would be the overlong encoding of the code point U+0 and you would have `c != '\0'` but `Int(c) == 0`. So if you're looking for a literal NUL byte, then comparing characters against `'\0'` will not accidentally find overlong encodings of U+0. If you're explicitly searching for characters with code point 0, then you will find it, however. That seems like a sane and useful distinction which allows people to work with common UTF-8 variants without getting into trouble.

* Split anything else into `Char` values according to the Unicode 10 standard. This is mainly so that we only ever have to do a single byte of look-ahead, instead of having to look ahead three bytes. I actually think that splitting everything else as single-byte invalid `Char`s is probably a more useful splitting, but this is easier to implement and it's closer to the standard, so  ¯\\\_(ツ)\_/¯.

This doesn't give 100% Unicode 10 conformance, but as I said above, I'm not entirely sure it really applies. The only observable discrepancy we might encounter is counting the number of characters in invalid strings differently, and in those cases, I think our answers will be what people actually expect/want. That seems like a pretty acceptable tradeoff in exchange for handling common UTF-8 variants smoothly.

bkamins commented:
First about reading 4-5 byte sequences. As I have shown in #24420 currently Julia does this, e.g. run `collect(String([0xff, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0, 0x41, 0x42, 0x43]))` (this is just for a reference).

I am OK with parsing valid overlong encoding and valid reserved surrogates as if they were allowed. And all other invalid UTF-8 following Unicode 10. I understand this is what you propose and this will lead to fastest implementation of splitting. We should simply document that we work like this

However, the other issue is conversion from `Char` to `UInt32`. In my bijection proposal above valid reserved surrogates would be decoded as if they were valid - exactly as you propose. However, if we want a bijection between `Char` and `UInt32` we cannot convert overlong encodings to their natural representation.

So the question is do we want to have this bijection or not.

And even if we do not want a full bijection I propose to decode all other than overlong and surrogate invalid `Char` to code points larger than U+10FFFF so that such invalid `Char` does not decode to a valid codepoint (and if we drop bijection we can apply some simple conversion rule not caring that they all all diffferent).

Apart from that then `isvalid(c::Char)` function should allow for *Unicode 10 compliant* checking and *Julia compliant* checking (if we end up with deciding that they do not have to be identical). Then probably some additional argument `strict::Bool` would be good.

StefanKarpinski commented:
I think I've changed my mind on the bijection business, after contemplating it. When converting from `Char` to integer seems to be the natural point to get an error if the `Char` does not naturally give you a code point. So I just need to make that conversion throw the error in a way that doesn't slow everything down so much. I think a more general character classification scheme may be better than the current `isvalid` function, which can only give two possible answers. `classify(c::Char)` and `classify(s::String)` or something like that. Could maybe be rolled in with the other `isalpha`, etc. stuff which we've kind of wanted to clean up [for some time now](https://github.com/JuliaLang/julia/issues/14347).

bkamins commented:
Maybe I am stating the obvious but I have just realized (maybe I am stating the obvious) that actually internally in Julia we will not need to convert from `Char` to `UInt32` at all.
One can work on `Char` that is UTF-8 encoded everywhere (i.e. even if some internal functions wanted Unicode parameter they can be rewritten to to take UTF-8 encoded parameter without making a conversion to a code point) - it might even be faster in some cases (e.g. `next` should be faster).
And in what cases is the exact code point value required by a programmer - I suppose very rarely - mostly in cases when some external library wants its input in Unicode.

StefanKarpinski commented:
Yes, that's exactly the idea – you can do most things with `Char` values without ever converting to code points. Actually wanting the code point for a character is quite rare! Why would you care? You can do almost everything by equality and inequality comparisons with other characters. And most of the time you don't even need to do that since you can just pass string data all the way through.

StefanKarpinski committed:
`Char` representation: UTF-8 bytes with least significant padding

StefanKarpinski committed:
`Char` representation: UTF-8 bytes with most significant padding

StefanKarpinski commented:
I'm puzzling through ordering here and thought maybe you have some thoughts, @bkamins. What ordering should the entire space of characters – valid and invalid – have? It's fairly common to write code like `'a' <= c <= 'z'` so it seems somewhat dangerous for invalid characters to ever have the property that `valid₁ ≤ invalid ≤ valid₂`. The other consideration is that we might want sorting lexicographically by character to match sorting lexicographically by bytes – which is a property that valid UTF-8 is designed to have. However, I don't think that's actually possible for any ordering of partial characters, so perhaps that shouldn't be a consideration after all. Any thoughts?

bkamins commented:
I think that we can order `Char` by bytes and inform users that:
* `Char` is UTF-8 encoded thus it is safe to write ranges only of ASCII characters (I would say that this is more or less the only case when character comparisons are made in practice);
* if they have only valid UTF-8 it will be ordered by code point (by valid I mean strictly valid - without overlong sequences);
* if they want ordering by code point and have no guarantee that UTF-8 is valid then they should convert `Char` to code point and compare it (and here I would say again that invalid UTF-8 should be always converted to some value larger than U+10FFFF) -> currently your code does not do that
* add a general remark that comparing non ASCII `Char` is not a good idea (so we are back to point 1 above) as in general right order of `Char` and `String` is locale dependent, e.g. in Polish 'a<ą<b<c<ć<d<e<ę<f`;

In short - outside of ASCII range I would not worry too much about comparing `Char` as there is no single valid rule anyway and it should be discouraged.

As a side comment - supporting https://en.wikipedia.org/wiki/Unicode_collation_algorithm somehow in Julia would be great.
In general I believe that if people will start using Julia for more text processing some package like [stringi](http://www.gagolewski.com/software/stringi/) in R will be crucial to have.
Are you aware of any such efforts?

StefanKarpinski commented:
I think we're in agreement here. The conclusion I came to is that the invariant we want is that `c₁ < c₂` if and only if `"$c₁" < "$c₂"`. This is naturally the case for valid UTF-8 lexically sorted by bytes, and extends to invalid characters as well by byte ordering – which is what I think you're saying too. That's a nice property that doesn't have to do with the representation of `Char`. However, it does imply that I haven't quite represented characters correctly here: the zero padding should be least significant, not most significant. I'll adjust the PR accordingly.

StefanKarpinski commented:
Also: I think that all fancy Unicode stuff like the collation algorithm can be implemented in packages.

bkamins commented:
The property you have indicated would be nice to have.

Two things to consider though:
* I have been thinking of right padded representation earlier and the problem I saw was that it would be more expensive to work with such `Char` than in current model. Re-implementing `convert`, `read` and `write` methods will show if there is a penalty.
* The property extends to `"$c₁[anything 1 here]" < "$c₂[anything 2 here]"` only for valid UTF-8. Below I am using right padding notation. Consider a valid character `0xF0908D88` which is greater than invalid character `0xF0908D00` that uses three bytes. Now compare the following 8-element sequences of bytes
`[0xF0, 0x90, 0x8D, 0x88, 0xF0, 0x90, 0x8D, 0x88]`
and
`[0xF0, 0x90, 0x8D, 0xF0, 0x90, 0x8D, 0x88, 0x00]`
the first one is less than than the second.
Though the first decoded to characters is
`[0xF0908D88, 0xF0908D88]`
and the second is
`[0xF0908D00, 0xF0908D88, 0x00000000]`

nalimilan commented:
Very nice! I agree that consistency between `c₁ < c₂` and `"$c₁" < "$c₂"` sounds essential.

Regarding over-long character encodings, I'd say it's fine to treat them as different from their correct counterparts for `<` as long as we do that consistently, i.e. they should *never* to be considered as equal, even after conversion to `UInt32`. Is that the plan?

bkamins commented:
Just to highlight the issue with comparison I see using analogous notation:
it is impossible to guarantee under any conversion rule that `s₁[1]<s₂[1]` and `s₁<s₂` are consistent if the latter uses `memcmp`.
Thus consistency between `c₁ < c₂` and `"$c₁" < "$c₂"` is a nice feature but will have a limited usefulness in practice.

The consistency `s₁[1]<s₂[1]` and `s₁<s₂` is guaranteed for valid UTF-8 strings no matter if we use character encoding with left or right padding - both will work.

So I would say that a major decision between left and right padding would be how they compare in performance.

As for conversion from `Char` to `UInt32`:
*  the current implementation in this PR maps overlong sequences to their code point representation (this not following what you have stated); also it maps invalid UTF-8 into valid code points;
* in the discussion above the idea was to make the conversion bijective (thus ensuring what you have said).
* the other approach would be to strictly encode all invalid UTF-8 sequences by U+FFFD and all valid UTF-8 to their code point value (this also would ensure that overlong sequences are not considered equal to their valid counterparts).

If I understand it correctly @StefanKarpinski wants to implement yet something else. Correct?

As another thought maybe it would be useful to define `CodePoint` type that would provide a different conversion behavior (i.e. strict conversion following Unicode 10 and always using U+FFFD to represent invalid UTF-8) than conversion to `UInt32` (which could be less strict)?

StefanKarpinski commented:
My thinking was that over-long encodings and surrogate pair code points would simply convert to their natural numerical code point values. That way, we naturally handle all kinds of UTF-8 variants without unwanted errors (and they are usually both unnecessary and unwanted), but you can still easily check for a non-standard encoding of a character, e.g. by checking that `Int(c) == 0` but `c != '\0'` – or more generally, `nonstandard_encoding(c::Char) = Char(Int32(c)) != c`. I had the following written down in some of my notes from yesterday:

> Let’s distinguish invalid code points and ill-formed characters. Invalid code points follow the structure of valid UTF-8 characters and therefore can be decoded to numerical code point values – but those code points are not valid according to the Unicode standard. An ill-formed character does not follow the UTF-8 encoding scheme and therefore cannot be decoded to a numerical code point at all.

Hopefully that distinction makes sense. For invalid code points that are still well-formed, we can decode them to a viable numerical value and I think we should do so because these often do occur (e.g. in Modified UTF-8 or WTF-8) and it's useful to be able to work with them in UTF-8-like data as code points. Ill-formed characters cannot even be sensibly decoded to a numerical code point value, so attempting to do so should raise an error.

bkamins commented:
@StefanKarpinski The rule is clear for me except one area.
How do you want to treat 4-byte correctly formed UTF-8 that decodes to code point larger than U+10FFFF.
This is the case when first byte has leading bits `11110` and three following bytes start with `10` bits but when decoded they are greater than `0x10ffff`. Their range is from `0xf4908080` to `0xf7b7bfbf`.
Do you want to decode them to integer or throw an error?

nalimilan commented:
Makes sense. I just wonder whether, to be completely safe, we shouldn't force people to use a special function (e.g. `codepoint`) when they want to convert a `Char` to a code point integer value. That way we could require setting an argument to explicitly allow invalid code points, rather than silently allowing them. This kind of detail is really tricky and can introduce security issues, so better be cautious and force people to think about what they are doing when they go beyond the basics.

bkamins commented:
@nalimilan Exactly because of these risks I have thought that your `codepoint` function could be a `CodePoint` type constructor so that you are forced to think what you are doing.

Another issue is to decide how `+` and `-` on `Char` should work (I assume we want to retain those functions).
I understand that the idea is that `Char` should be converted to code point, incremented/decremented and converted back to `Char`. Yes? If this is the case it should be made clear what do we want to happen when the starting `Char` is invalid or when the result of addition is greater than `0x0010FFFF`.

However, whatever we do it should be expected that `+` and `-` operations on `Char` will have a performance regression in comparison to the current implementation of `Char`.

stevengj commented:
`UInt32(::Char)` or `convert(UInt32, ::Char)` to get the codepoint seems perfectly comprehensible to me (for valid characters): the code point seems like the only reasonable conversion to me, so I'm not in favor of a `codepoint` function.  Having to call a function here seems like an annoyance.

You can always do `reinterpret(UInt32, Char)` if you want the raw internal representation.

stevengj commented on base/char.jl:
```diff
@@ -29,11 +45,9 @@ done(c::Char, state) = state
-hash(x::Char, h::UInt) = hash_uint64(((UInt64(x)+hashchar_seed)<<32) ⊻ UInt64(h))
+==(x::Char, y::Char) = reinterpret(UInt32, x) == reinterpret(UInt32, y)
+isless(x::Char, y::Char) = reinterpret(UInt32, x) < reinterpret(UInt32, y)
+hash(x::Char, h::UInt) = hash(reinterpret(UInt32, x), hash(Char, h))
```
According to `BenchmarkTools` this is 5 times slower than the old `hash(x::Char)`.  Should we use 
```jl
hash(x::Char, h::UInt) = hash_uint64(((reinterpret(UInt32, x)+UInt64(hashchar_seed))<<32) ⊻ UInt64(h))
```
or similar instead?

StefanKarpinski replied:
Perhaps. I worry about that xor being symmetric. I also worry that we're using this kind of pattern all over Base and it is quite inefficient. We really need a better way to express this.

StefanKarpinski replied:
Went with your suggested definition for now. If there's some hash collision issue with the existing definition, that's a totally independent issue to this representation change so can be addressed separately.

stevengj commented on base/filesystem.jl:
```diff
@@ -149,6 +149,26 @@ function read(f::File, ::Type{UInt8})
     return ret % UInt8
 end
 
+function read(f::File, ::Type{Char})
```
Why is `function read(s::IO, ::Type{Char})` not sufficient here?

StefanKarpinski replied:
This is annoying: `IO` doesn't support `position` and `seek`; `File` doesn't support `mark` and `reset`. That could be fixed instead, but this was the easiest way to get this working.

stevengj commented on base/stream.jl:
```diff
@@ -1148,6 +1148,14 @@ unmark(x::LibuvStream)   = unmark(x.buffer)
 reset(x::LibuvStream)    = reset(x.buffer)
 ismarked(x::LibuvStream) = ismarked(x.buffer)
 
+function peek(s::LibuvStream)
```
Why is `peek(s::IO)` not sufficient here?

StefanKarpinski replied:
`LibuvStream` isn't defined when that method is defined; I don't recall if I tried making the definition here. In general, this IO stuff is a bit of a mess because we have Julia objects, Libuv objects, flisp/support objects, and raw system objects. This all needs to be significantly simplified.

nalimilan commented:
> UInt32(::Char) or convert(UInt32, ::Char) to get the codepoint seems perfectly comprehensible to me (for valid characters): the code point seems like the only reasonable conversion to me, so I'm not in favor of a codepoint function. Having to call a function here seems like an annoyance.

So you suggest throwing an error for invalid code points? Works for me, but at some point we might need to introduce a dedicated function anyway with an argument to allow invalid code points.

catawbasam commented:
@bkamins -- re  stringi , have you looked over ICU.jl?  That might provide a starting point.

StefanKarpinski commented:
My general thinking is that the philosophy of `String` should be that it is a UTF-8 like interpretation of some bytes and that if there is a reasonable interpretation of something, it should work without errors or other fussiness. If someone wants a strict UTF-8 string type, that can and should be provided by an external package, but the most common and standard use case is not so strict and doing something “reasonable” is the right default behavior.

Converting a `Char` to `UInt32` should, IMO work for any well-formed UTF-8 character even if the code point is invalid. We can extend `isvalid(c::Char)` to tell you if it's a valid Unicode character or not. I'm not certain what to do about converting the the U+110000 to U+1f7fff range. On the one hand, those aren't valid Unicode code points; less so, even than invalid surrogate pair code points, which "exist" but aren't supposed to be used by themselves. On the other hand, considering the above philosophy, what other numeric value could they produce? So why shouldn't we produce those numercial value when asked for them?

Another option is that we simply shouldn’t produce those `Char` values when iterating string data. In other words, when you encounter the bytes 0xf4, 0x90 you should already just return 0xf4 as an invalid `Char` by itself, so no `Char` values that would decode as anything from U+110000 to U+1f7fff would exist in the first place (at least by using normal APIs rather than creating `Char` values directly). Perhaps I’ve answered my own question here: don’t generate `Char` values for U+110000 to U+1f7fff, but if you do have a `Char` value in the that range, just decode it according to normal UTF-8 rules.

nalimilan commented:
Actually, I'm fine with allowing invalid code points, i.e. those beyond U+10FFFF. Indeed that's required to work with possibly invalid Windows file paths AFAIK (WTF-8).

What I'm concerned about is the handling of overlong encodings, and specifically that of the null code point à la Modified UTF-8 (`0xc080`). Failing to catch a null codepoint when checking a string can often create security breaches. If we allow that sequence to produce a `Char` when iterating over a string, I think having (as currently) `String(UInt8[0xc0, 0x80])[1] == '\0'` is in order. As you suggest, we could also avoid creating such `Char`s in the first place. But I don't really like the idea that some `Char` values would be possible, but not used by the default implementation: what if a custom implementation creates them (on purpose or by mistake)? Most code won't be prepared to handle them.

StefanKarpinski commented:
I don't really understand what the security issue is. Julia doesn't care about NUL bytes in strings – it's just data. Is the concern that Julia code will look for `c == 0`, find an overlong `\0` which nevertheless has code point zero, conclude that the data is correctly terminated and then pass it to a C function which looks for the terminating NUL byte and doesn't find it, thereby wandering off into invalid memory? If so, there's a couple of things we can do to prevent that:

1. Tell people to check for null bytes by doing `c == \0` rather than `c == 0`.
2. Tell people to pass strings to C APIs using the `Cstring` type which checks that there are no embedded NUL bytes inside the string and that the string data is NUL terminated.

nalimilan commented:
Yes, I'm fine with your two suggestions. What I wanted to ensure is that `c == '\0'` would catch overlong `'\xc0\x80'` characters (it wasn't clear to me that it would be the case in your proposal), since that's the most natural way of checking for code points.

StefanKarpinski commented:
Real use case for high code points – code like this:

https://github.com/JuliaLang/julia/blob/c80df3cd5328be68c5dd2b5f549504d79e7c8d0c/base/strings/io.jl#L324-L326

Without parentheses, `n<<4 + c` can end up being a too large code point which would throw an error upon construction, but `n<<4 + c-'a'+10` ends up being a valid code point. This could be fixed by reassociating the operations, but that seems annoyingly fussy. Of course, reassociating these operations also makes this much more efficient with the new character representation.

StefanKarpinski committed:
iswellformed(c::Char) to test if `c` represents a code point

This allows testing before attempting to decode `c` which we use
in a number of places to handle malformed UTF-8 data correctly.

StefanKarpinski committed:
strings: some cosmetic tweaks

StefanKarpinski committed:
remove internal chomp! function

this function was always dangerous and has been equivalent to `chomp`
since the `String` representation change introduced in Julia 0.6

StefanKarpinski committed:
convert(String, ::Vector{Char}): don't normalize surrogate pairs

we're in a new era where Julia tries as hard as it can to pass text
data through intact rather than trying to guess what you meant. to
that end, when converting a vector of characters to a `String` we
no longer detect and combine surrogate pairs. there can be a utility
function to do that kind of normalization on strings, but it's not
the job of the basic string code.

StefanKarpinski committed:
delete unused `unescape_chars` function

StefanKarpinski committed:
malformed chars are always grapheme breaks

StefanKarpinski committed:
iswellformed => !ismalformed; make test stricter

despite generally trying to avoid double negatives, it seems that
"malformed" is the clearer and more common concept to test for here.
it was also a bit confusing to have `MalformedCharError` but then
have `iswellformed` as the predicate instead of `ismalformed`.

testing for well-formedness more strictly doesn't seem to add too
much complexity here; if performance becomes an issue, we'll need a
`@inbounds`-like annotation anyway.

StefanKarpinski committed:
slightly more efficient character checking and decoding

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip [ci skip]

stevengj commented:
@ScottPJones suggested that we have `Char <: AbstractChar` to allow for different character encodings to use different representations, which seems reasonable for future-proofing.  Not sure how much this would complicate Base, but at first glance it seems like it would be just changing a bunch of `Char` functions to accept `AbstractChar`.

StefanKarpinski commented:
Sure that seems reasonable. It's a fairly orthogonal change to this one.

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
Revert "wip"

This reverts commit bc7f25d83165fe4be209d6485406604c762930dd.

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
my dirty laundry, you filthy voyeur [ci skip]

StefanKarpinski committed:
wip [ci skip]

StefanKarpinski committed:
wip: added a doc string for AbstractString

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

StefanKarpinski committed:
wip

yurivish commented on base/regex.jl:
```diff
@@ -303,8 +303,12 @@ struct SubstitutionString{T<:AbstractString} <: AbstractString
-next(s::SubstitutionString, idx::Int) = next(s.string, idx)
+ncodeunits(s::SubstitutionString) = ncodeunits(s.string)
+codeunit(s::SubstitutionString) = codeunit(s.string)
+codeuint(s::SubstitutionString, i::Integer) = codeunit(s.string, i)
```
Should this be `codeunit` rather than `codeuint`?

StefanKarpinski replied:
why yes, yes it should. thank you!

StefanKarpinski committed:
fix [ci skip]

yurivish commented on base/strings/basic.jl:
```diff
@@ -1,57 +1,188 @@
-The number of bytes in string `s`.
+    codeunit(s, i) :: codeunit(s)
+
+I.e. the value returned by `codeunit(s, i)` is of the type returend by
```
`returned`? 😄

StefanKarpinski committed:
docstring typo fix [ci skip]

StefanKarpinski committed:
test for more method errors [ci skip]

StefanKarpinski commented:
> Does this need a test?

Probably. Lots of things need tests.

StefanKarpinski committed:
cosmetic tweaks

StefanKarpinski committed:
wip
